src/common/cockpitjson.c

```diff
@@ -967,3 +967,165 @@ cockpit_json_to_hash_table (JsonObject *object,
 
   return hash_table;
 }
+
+typedef const struct {
+  CockpitJsonWalkCallback callback;
+  gpointer                user_data;
+} WalkClosure;
+
+static JsonNode *
+cockpit_json_walk_node (JsonNode    *node,
+                        WalkClosure *closure);
+
+static JsonArray *
+cockpit_json_walk_array (JsonArray   *array,
+                         WalkClosure *closure)
+{
+  gint n = json_array_get_length (array);
+  JsonArray *copy = NULL;
+
+  for (gint i = 0; i < n; i++)
+    {
+      JsonNode *element = json_array_get_element (array, i);
+      JsonNode *node = cockpit_json_walk_node (element, closure);
+
+      if (node && !copy)
+        {
+          /* We've decided we need to copy now, so revisit everything up
+           * to this point and add it to the copy.
+           */
+          copy = json_array_sized_new (n);
+          for (gint j = 0; j < i; j++)
+            json_array_add_element (copy, json_array_dup_element (array, j));
+        }
+
+      if (copy)
+        json_array_add_element (copy, node ?: json_node_ref (element));
+    }
+
+  return copy;
+}
+
+static JsonObject *
+cockpit_json_walk_object (JsonObject  *object,
+                          WalkClosure *closure)
+{
+  JsonObject *copy = NULL;
+  JsonObjectIter iter;
+  const gchar *key;
+  JsonNode *value;
+
+  json_object_iter_init (&iter, object);
+  while (json_object_iter_next (&iter, &key, &value))
+    {
+      JsonNode *node = cockpit_json_walk_node (value, closure);
+
+      if (node && !copy)
+        {
+          const gchar *this_key = key;
+
+          /* We've decided we need to copy now, so revisit everything up
+           * to this point and add it to the copy.
+           */
+          copy = json_object_new ();
+          json_object_iter_init (&iter, object);
+          while (json_object_iter_next (&iter, &key, &value) && key != this_key)
```

Comment:
I can't imagine that this wouldn't be the case, particularly since this object is immutable.

The underlying object is a hash table, of course, and this iterator is indeed backed by a `GHashTableIter`, which has no guarantees of any kind on order, but won't change if nothing in the table changes.

I have the same concerns for hashing and equality comparisons, but again...

Reply:
Ack, thanks. Hopefully our tests would catch if this is ever not the case, but I agree it's rather unlikely to not be a reproducible order.